specs/core/0_beacon-chain.md

```diff
@@ -644,6 +644,39 @@ def is_slashable_attestation_data(data_1: AttestationData, data_2: AttestationDa
     )
 ```
 
+#### `is_valid_indexed_attestation`
+
+```python
+def is_valid_indexed_attestation(state: BeaconState, indexed_attestation: IndexedAttestation) -> None:
```

Comment:
Now we need to make it return `-> bool`, and refactor to use `return`. My favourite is something like

```python
return (
    len(bit_1_indices) == 0 and
    len(bit_0_indices) + len(bit_1_indices) <= MAX_VALIDATORS_PER_COMMITTEE and
    len(set(bit_0_indices).intersection(bit_1_indices)) == 0 and
    bit_0_indices == sorted(bit_0_indices) and bit_1_indices == sorted(bit_1_indices) and
    bls_verify_multiple(
        pubkeys=[
            bls_aggregate_pubkeys([state.validators[i].pubkey for i in bit_0_indices]),
            bls_aggregate_pubkeys([state.validators[i].pubkey for i in bit_1_indices]),
        ],
        message_hashes=[
            hash_tree_root(AttestationDataAndCustodyBit(data=indexed_attestation.data, custody_bit=0b0)),
            hash_tree_root(AttestationDataAndCustodyBit(data=indexed_attestation.data, custody_bit=0b1)),
        ],
        signature=indexed_attestation.signature,
        domain=get_domain(state, DOMAIN_ATTESTATION, indexed_attestation.data.target.epoch),
    )
)
```

Reply:
Alternatively we could just put `return True` at the end and call it a day

Reply:
On it

Reply:
We made this into a `validate` with separated statements to allow for ease in debugging. After surveying implementers, they greatly preferred _not _ grouping this all into one boolean.

Prefer sticking with `validate`.

Reply:
Lol.

Reply:
What about Justin's many-asserts-followed-by-a-return option? It is less pretty from a code standpoint, but conceptually I think this function should be a predicate.

Reply:
Not seeing the issue with a `validate_` function handling asserts.

The alternative is either (1) overly verbose but having if/return statements for each condition or (2) semantically  strange if we have asserts and then a return True.

Reply:
We debated this a month ago and was the compromise we (including @vbuterin) agreed upon

Reply:
cc: @JustinDrake

Reply:
> overly verbose but having if/return statements for each condition 

I considered this momentarily before proposing going back to the many asserts then return option.

My opinions aren't really that strong about this. If there was a discussion on this topic and the conclusion was a preference for the original, I will revert.

> 2) semantically strange if we have asserts and then a return True.

Agreed this is strange. It is a lot of ugliness just to so a function name is slightly more aligned with others in the spec.

Reply:
> We debated this a month ago and was the compromise we (including @vbuterin) agreed upon

Yes this was debated. I guess there are new circumstances now, given how `is_valid_indexed_attestation` makes all the helpers fall into nice clean categories.

Reply:
I'm against collapsing into one boolean (esp as someone who writes and maintains a bunch of tests). 

The `if not (condition): return False` is verbose but seems like where the compromise currently lies

Reply:
I'm happy with the compromise :)